src/command/lsp/mod.rs

```diff
@@ -0,0 +1,230 @@
+use crate::composition::events::CompositionEvent;
+use crate::composition::runner::Runner;
+use crate::composition::supergraph::binary::OutputTarget;
+use crate::composition::supergraph::config::{
+    ResolveSupergraphConfigError, SupergraphConfigResolver,
+};
+use crate::composition::supergraph::install::InstallSupergraph;
+use crate::composition::{
+    CompositionError, CompositionSubgraphAdded, CompositionSubgraphRemoved, CompositionSuccess,
+};
+use crate::utils::effect::exec::TokioCommand;
+use crate::utils::effect::install::InstallBinary;
+use crate::utils::effect::read_file::FsReadFile;
+use crate::utils::effect::write_file::FsWriteFile;
+use crate::{
+    options::PluginOpts,
+    utils::{client::StudioClientConfig, parsers::FileDescriptorType},
+    RoverOutput, RoverResult,
+};
+use anyhow::{anyhow, Error};
+use apollo_federation_types::config::FederationVersion;
+use apollo_language_server::{ApolloLanguageServer, Config};
+use camino::Utf8PathBuf;
+use clap::Parser;
+use futures::StreamExt;
+use serde::Serialize;
+use std::collections::HashMap;
+use std::env::temp_dir;
+use std::io::stdin;
+use tower_lsp::lsp_types::{Diagnostic, Range};
+use tower_lsp::Server;
+use tracing::debug;
+
+#[derive(Debug, Serialize, Parser)]
+pub struct Lsp {
+    #[clap(flatten)]
+    pub(crate) opts: LspOpts,
+}
+
+#[derive(Clone, Debug, Serialize, Parser)]
+pub struct LspOpts {
+    #[clap(flatten)]
+    pub plugin_opts: PluginOpts,
+
+    /// The absolute path to the supergraph configuration file.
+    #[serde(skip_serializing)]
+    #[arg(long = "supergraph-config")]
+    supergraph_yaml: Option<Utf8PathBuf>,
+}
+
+impl Lsp {
+    pub async fn run(&self, client_config: StudioClientConfig) -> RoverResult<RoverOutput> {
+        self.opts
+            .plugin_opts
+            .prompt_for_license_accept(&client_config)?;
+
+        run_lsp(client_config, self.opts.clone()).await?;
+        Ok(RoverOutput::EmptySuccess)
+    }
+}
+
+async fn run_lsp(client_config: StudioClientConfig, lsp_opts: LspOpts) -> RoverResult<()> {
+    //TODO: Check this error handling is right i.e. if we don't get a supergraph.yaml passed
+    // this should fail rather than doing something else.
```

Comment:
This one I wasn't particularly sure about, in my view, with this command, we should make passing a Supergraph YAML compulsory because I can't really understand what not passing one would even mean. That would mean we could make `clap` take it as a Path, rather than an Option and tidy this code up a bit.

Reply:
The language server does standard GraphQL validations all on its own, outside of federation. So you can totally use our extension with `rover lsp` for a monograph if you don't pass in a `supergraph.yaml` file. Passing in that file is what triggers the composition mode—so most of this logic. If you don't pass in the file, `rover lsp` is mostly just a passthrough to the language server crate.

Reply:
@dylan-apollo In that case, suppose we don't get a `supergraph.yaml` passed, what value(s) should be sent to the `ApolloLanguageServer::build_service` function? 

How do you know what graph you're supposed to be targetting?

Reply:
It should already work as-is, the code on my branch collected into an empty hashmap for the subgraphs and then just didn't spawn the composition worker.

> How do you know what graph you're supposed to be targetting?

I don't understand this question

Reply:
Sorry, that's me being dim, I apologise, I've made the change now so that if we don't get a `supergraph.yaml` we bail out early before doing any composition etc. and then everything else can continue as is.